### PR TITLE
[dagit-bb] Tabs component

### PR DESCRIPTION
**Pull request by bengotow**
I originally planned to reskin the Blueprint components for this one, but we aren't using much of the Blueprint component surface area and a fresh implementation for the callsites in Dagit was only ~100loc. Don't have a strong preference but this seemed like it'd be a bit easier to maintain.

It might be interesting to try to align how the icon is passed in here with the Tags PR, I put the color on the icon but we could put iconColor on the Tab to match instead.

Sep-20-2021 14-11-43

## Test Plan
Try out new Tabs storybook

## Checklist
- [ ] My change requires a change to the documentation and I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.

### CODE REVIEW

File: js_modules/dagit/packages/core/src/ui/Tabs.tsx [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  children: Array<React.ReactElement<TabProps>>;
+  selectedTabId?: string;
+  onChange?: (selectedTabId: string) => void;
+  small?: boolean;
+}
+
+export const Tabs = styled(({selectedTabId, children, small, onChange, ...rest}) => {
+  if (!(children instanceof Array) || children.some((c) => c.type !== Tab)) {
+    throw new Error('Tabs must render Tab instances');
+  }
```
Comment (hellendag) [inline js_modules/dagit/packages/core/src/ui/Tabs.tsx:18]: Is TS enforcement insufficient on this?

File: js_modules/dagit/packages/core/src/ui/Tabs.tsx [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+import {ColorsWIP} from './Colors';
+import {IconWrapper} from './Icon';
+import {FontFamily} from './styles';
+
+interface TabsProps {
+  children: Array<React.ReactElement<TabProps>>;
+  selectedTabId?: string;
+  onChange?: (selectedTabId: string) => void;
+  small?: boolean;
```
Comment (hellendag) [inline js_modules/dagit/packages/core/src/ui/Tabs.tsx:12]: Minor, but I kind of prefer `size: 'small' | 'large';` with a default.

I always find it confusing when some components have `small: boolean` and others have `large: boolean`.

File: js_modules/dagit/packages/core/src/ui/Tabs.tsx [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+export const Tabs = styled(({selectedTabId, children, small, onChange, ...rest}) => {
+  if (!(children instanceof Array) || children.some((c) => c.type !== Tab)) {
+    throw new Error('Tabs must render Tab instances');
+  }
+
+  return (
+    <div {...rest} role="tablist">
+      {children.map((child) =>
+        React.cloneElement(child, {
```
Comment (hellendag) [inline js_modules/dagit/packages/core/src/ui/Tabs.tsx:23]: Safer to use `React.Children.map`?

**Review by hellendag — APPROVED** [review]

File: js_modules/dagit/packages/core/src/ui/Tabs.tsx (hunk 1 above)
Comment (bengotow) [reply js_modules/dagit/packages/core/src/ui/Tabs.tsx:18]: Unfortunately it seems so... I'll poke at this some more but I haven't be en able to get a TS error on other types of component children, only on basic primitives like string children.

image

File: js_modules/dagit/packages/core/src/ui/Tabs.tsx (hunk 3 above)
Comment (bengotow) [reply js_modules/dagit/packages/core/src/ui/Tabs.tsx:23]: Oh interesting I guess so - I hadn't seen that before. It looks like I still need this cloneElement line though?

**Pull request merged**